cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,381 @@
+   *                             indicated by `left_on[i]`.
+   * @param[in] return_columns   A vector of column indices from `left` to
+   *                             include in the returned table.
+   * @param[in] mr               Device memory resource to use for device memory allocation
+   *
+   * @returns                    Result of joining `left` and `right` tables on the columns
+   *                             specified by `left_on` and `right_on`. The resulting table
+   *                             will contain `return_columns` from `left` that match in right.
+   */
+  std::unique_ptr<cudf::experimental::table> left_anti_join(cudf::table_view const& left,
+                                                            cudf::table_view const& right,
+                                                            std::vector<cudf::size_type> const& left_on,
+                                                            std::vector<cudf::size_type> const& right_on,
+                                                            std::vector<cudf::size_type> const& return_columns,
+                                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  /**
+   * @brief Type defining join comparison operators
+   *
+   * Indicates how two elements `a` and `b` compare with one and another.
+   */
+  enum class join_comparison_operator {
+    LESS_THAN,             ///< Compare a < b
+    LESS_THAN_OR_EQUAL,    ///< Compare a <= b
+    EQUAL,                 ///< Compare a == b
+    GREATER_THAN,          ///< Compare a >  b
+    GREATER_THAN_OR_EQUAL, ///< Compare a >= b
+  };
+
+  /**
+   * @brief Type defining non-equi join operations
+   */
+  struct join_operation {
+    join_comparison_operator op;                   ///< type of comparison
+    cudf::size_type          left_column_idx;      ///< index of left column to compare
+    cudf::size_type          right_column_idx;     ///< index of right column to compare
+    bool                     only_left_in_output;  ///< if true, only output the left column
+  }
+  
+  /** 
+   * @brief  Performs a left non-equi inner join on two tables (left, right)
+   *
+   * A left non-equi inner join acts like an inner join, however it allows
+   * for comparison conditions beyond equality.
+   *
+   * Ultimately, this interface could be quite complex.  For the moment it is limited
+   * to defining a list of conditions that will be anded together to provide the join
+   * criteria.
+   *
+   * @example TableA a: {0, 1, 2}
+   *          TableB b: {1, 2, 3}, a: {1, 2, 5}
+   *          join_ops: { {LESS_THAN, 0, 1, true} }
+   * Result: { a: {1, 2, 5}, b: {1, 2, 3} }
+   *
+   * @example TableA a: {0, 1, 2}, c: {1, 2, 5}
+   *          TableB b: {1, 2, 3}
+   *          join_ops: { {GREATER_THAN, 0, 0, false} }
+   * Result: { a: {2}, b: {1}, c: {5} }
+   *
+   * @throws cudf::logic_error if either table is empty
```

Comment:
Seems like there are some other possibilities here:   Out of bounds indices in columns_in_common, out of bounds indices in the join_ops structs.